page_analyzer/utils/db.py

```diff
@@ -68,10 +68,66 @@ def create_url(url_name: str) -> int:
     return int(url_id)
 
 
-def get_urls_data() -> List[int | str]:
+def get_urls_data() -> List[Dict["str", Any]]:
+    conn = connect_db()
+    with conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as dict_cur:
+        dict_cur.execute(
+            "SELECT"
+            " urls.id AS url_id,"
+            " urls.name AS url_name,"
+            " last_checks.created_at AS date,"
+            " last_checks.status_code AS response_code"
+            " FROM urls"
+            " LEFT JOIN"
+            " (SELECT DISTINCT ON (url_id) url_id, created_at, status_code"
+            " FROM checks"
+            " ORDER BY url_id, created_at, status_code DESC)"
+            " AS last_checks"
+            " ON last_checks.url_id=urls.id"
+            " GROUP BY"
+            " urls.id,"
+            " urls.name,"
+            " date,"
+            " response_code"
+            " ORDER BY urls.id DESC;"
```

Comment:
can be simplified to smth like this (post - `"parent" table, comments - "child" table):

```
select distinct on (p.id) p.* c.*
from posts p left join
     comments c
     on p.id = c.post_id
order by p.id desc, c.created_at desc;
```

Reply:
let's try this query

```sql
select distinct on (u.id) u.id as url_id, u.name as url_name, c.created_at as check_created_at
from urls u
left join checks c on c.url_id = u.id
order by u.id desc, c.created_at desc; 
```